ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java

```diff
@@ -287,16 +302,27 @@ private void buildAddClauseForAlter(StringBuilder query) {
   private void buildSelectClauseForInsert(StringBuilder query) {
     // Need list of columns for major crud, mmmajor partitioned, mmminor
     List<FieldSchema> cols;
-    if (major && crud || major && insertOnly && sourcePartition != null || minor && insertOnly) {
+    if (rebalance || major && crud || major && insertOnly && sourcePartition != null || minor && insertOnly) {
       if (sourceTab == null) {
         return; // avoid NPEs, don't throw an exception but skip this part of the query
       }
       cols = sourceTab.getSd().getCols();
     } else {
       cols = null;
     }
-
-    if (crud) {
+    if (rebalance) {
+      query.append("0, t2.writeId, t2.rowId % ");
```

Comment:
What is the meaning of this "0"? Does it have some significance here?

Reply:
The first value id the 'operation'. In other compactions it is set by the validate_acid_sort_order UDF, but in the rebalancing case that function would be a more difficult to use.